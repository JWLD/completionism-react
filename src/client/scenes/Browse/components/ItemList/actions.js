import * as ACTIONS from 'constants/action_types'

export const openInfoPanel = () => ({
  type: ACTIONS.OPEN_INFO_PANEL
})

export const setActiveItemId = itemId => ({
  type: ACTIONS.SET_ACTIVE_ITEM_ID,
  payload: itemId
})

export const setActiveItem = itemId => dispatch => {
  dispatch(setActiveItemId(itemId))
  dispatch(openInfoPanel())
}
