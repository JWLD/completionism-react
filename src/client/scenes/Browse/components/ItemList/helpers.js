import sortBy from 'lodash/sortBy'

import { SOURCES } from 'constants/sources'

// FILTERING HELPERS

export const checkFaction = (itemFaction, faction) => {
  const factionIsNeutral = itemFaction === 2

  return itemFaction === faction || factionIsNeutral
}

export const checkFilterValue = (itemName, filterValue) => {
  const normalisedName = itemName.toLowerCase()
  const normalisedFilterValue = filterValue.toLowerCase()

  return normalisedName.includes(normalisedFilterValue)
}

// ORGANISING HELPERS

const organiseIntoSubCategoryBlocks = ({ items, source }) => {
  const sourceItems = items.filter(item => item.source === source)

  const subCategoryBlocks = sourceItems.reduce((acc, item) => {
    const blockExists = acc.find(block => block.name === item.sub1)

    if (!blockExists) {
      const subCategoryItems = items.filter(foo => foo.sub1 === item.sub1)

      acc.push({
        name: item.sub1,
        items: subCategoryItems
      })
    }

    return acc
  }, [])

  const sortedBlocks = sortItemBlocks(subCategoryBlocks)

  return sortedBlocks
}

export const organiseIntoSourceBlocks = items => {
  const sourceBlocks = items.reduce((acc, item) => {
    const sourceType = SOURCES[item.source]
    const blockExists = acc.find(block => block.name === sourceType)

    if (!blockExists) {
      acc.push({
        name: sourceType,
        items: organiseIntoSubCategoryBlocks({ items, source: item.source })
      })
    }

    return acc
  }, [])

  const sortedBlocks = sortItemBlocks(sourceBlocks)

  return sortedBlocks
}

// SORTING HELPERS

export const sortItemBlocks = data => {
  return sortBy(data, 'name')
}

export const sortNonPets = data => {
  return sortBy(data, ['skill', 'quality', 'name'])
}

export const sortPets = data => {
  return sortBy(data, ['name'])
}
