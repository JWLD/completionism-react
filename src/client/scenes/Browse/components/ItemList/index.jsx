import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import { dataSelector } from './selectors'

import * as SC from './styled'
import ItemTile from 'scenes/Browse/components/ItemTile/ItemTile'

const ItemList = props => {
  const renderBlockTiles = data => {
    return data.map(item => <ItemTile {...item} key={item.id} />)
  }

  const renderListBlocks = () => {
    return Object.keys(props.data).map(subCategory => {
      const subCategoryData = props.data[subCategory]

      return (
        <SC.ItemListBlock key={subCategory}>
          <SC.BlockTitle>{subCategory}</SC.BlockTitle>
          {renderBlockTiles(subCategoryData)}
        </SC.ItemListBlock>
      )
    })
  }

  return renderListBlocks()
}

ItemList.propTypes = {
  data: PropTypes.object.isRequired
}

const mapStateToProps = (state, ownProps) => ({
  data: dataSelector(state, ownProps)
})

export default withRouter(connect(mapStateToProps)(ItemList))
