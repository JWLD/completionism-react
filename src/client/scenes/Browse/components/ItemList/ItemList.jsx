import React from 'react'
import PropTypes from 'prop-types'

import ProgressBar from 'components/ProgressBar/ProgressBar'
import Spinner from 'components/Spinner/Spinner'

import ItemTile from '../ItemTile/ItemTile'
import {
  filterByField,
  filterByFaction,
  orderByFields,
  orderObjectByKeys,
  organiseIntoSubcats
} from '../../services/dataHelpers'
import * as SC from './styled'

const filterData = props => {
  let filtered = props.categoryData

  filtered = filtered.filter(item =>
    item.name.toLowerCase().includes(props.filterVal)
  )
  filtered = filterByField(filtered, 'content', props.content)
  filtered = orderByFields(filtered, ['name', 'quality'])

  const { faction } = localStorage[props.category]
    ? JSON.parse(localStorage[props.category]).char
    : 2
  filtered = filterByFaction(filtered, faction)

  return filtered
}

const organiseData = data => {
  let organised = organiseIntoSubcats(data)
  organised = orderObjectByKeys(organised)

  return organised
}

const createItemLists = (category, categoryData, data, storageData) => {
  if (data.length === 0) {
    return <Spinner size={5} />
  }

  if (Object.keys(data).length === 0) {
    return (
      <SC.ItemListBlock>
        <SC.BlockTitle>No Items</SC.BlockTitle>
      </SC.ItemListBlock>
    )
  }

  return Object.keys(data).map(subCat => {
    const subCatHeader = <SC.BlockTitle>{subCat}</SC.BlockTitle>

    const tiles = data[subCat].map(item => (
      <ItemTile
        key={item.id}
        {...item}
        category={category}
        categoryData={categoryData}
        storageData={storageData}
      />
    ))

    return (
      <SC.ItemListBlock key={subCat}>
        {subCatHeader}
        {tiles}
      </SC.ItemListBlock>
    )
  })
}

const ItemList = props => {
  const storageData = localStorage[props.category]
    ? JSON.parse(localStorage[props.category]).ids
    : []
  const filteredData = filterData(props)
  const organisedData = organiseData(filteredData)
  const itemLists = createItemLists(
    props.category,
    props.categoryData,
    organisedData,
    storageData
  )

  return (
    <SC.ItemList>
      <ProgressBar
        data={filteredData}
        storageData={storageData}
        routeProps={props.routeProps}
      />
      {itemLists}
    </SC.ItemList>
  )
}

ItemList.propTypes = {
  category: PropTypes.string.isRequired,
  categoryData: PropTypes.array.isRequired,
  routeProps: PropTypes.object.isRequired
}

export default ItemList
