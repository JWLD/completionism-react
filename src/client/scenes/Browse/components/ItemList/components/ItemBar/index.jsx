import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { ICON_URLS } from 'constants/urls'
import { setActiveItem } from 'ItemList/actions'
import { activeCategorySelector } from 'Browse/selectors'

import * as SC from 'ItemBar/styled'

export const ItemBar = ({
  category,
  collected,
  icon,
  id,
  level,
  name,
  quality,
  setActiveItem
}) => {
  const iconUrl = icon ? `url(${ICON_URLS.large}${icon}.jpg)` : 'none'

  const renderProgressIcon = () => {
    if (collected && category === 'pets') {
      const maxPet = level === 0 || (quality === 3 && level === 25)
      const petLevel = level === 0 ? '-' : level

      return <SC.PetLevel maxPet={maxPet}>{petLevel}</SC.PetLevel>
    } else if (collected) {
      return <SC.CheckIcon />
    } else {
      return <SC.CrossIcon />
    }
  }

  return (
    <SC.ItemBar onClick={() => setActiveItem(id)}>
      <SC.ItemIcon iconUrl={iconUrl} />
      <SC.ItemTitle quality={quality}>{name}</SC.ItemTitle>
      {renderProgressIcon()}
    </SC.ItemBar>
  )
}

ItemBar.propTypes = {
  category: PropTypes.string.isRequired,
  collected: PropTypes.bool.isRequired,
  icon: PropTypes.string,
  id: PropTypes.number.isRequired,
  level: PropTypes.number,
  name: PropTypes.string.isRequired,
  quality: PropTypes.number.isRequired,
  setActiveItem: PropTypes.func.isRequired
}

ItemBar.defaultProps = {
  icon: null,
  level: 0
}

const mapStateToProps = state => ({
  category: activeCategorySelector(state)
})

const mapDispatchToProps = {
  setActiveItem
}

export default connect(mapStateToProps, mapDispatchToProps)(ItemBar)
