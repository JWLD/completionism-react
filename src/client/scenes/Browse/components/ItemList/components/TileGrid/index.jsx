import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { ICON_URLS } from 'constants/urls'
import { setActiveItem } from 'ItemList/actions'
import { activeCategorySelector } from 'Browse/selectors'

import * as SC from 'TileGrid/styled'

const TileGrid = ({ category, items, setActiveItem }) => {
  const tiles = items.map(({ collected, icon, id, level, quality }) => {
    const iconUrl = icon ? `url(${ICON_URLS.large}/${icon}.jpg)` : 'none'

    const renderQualityBar = () => {
      const petQuality =
        level === 0 || (quality === 3 && level === 25) ? 7 : quality

      return <SC.QualityBar petQuality={petQuality} />
    }

    return (
      <SC.TileWrap key={id}>
        <SC.Tile
          collected={collected}
          onClick={() => setActiveItem(id)}
          quality={quality}>
          <SC.TileIcon iconUrl={iconUrl} />
        </SC.Tile>
        {category === 'pets' && renderQualityBar()}
      </SC.TileWrap>
    )
  })

  return <SC.TileGrid>{tiles}</SC.TileGrid>
}

TileGrid.propTypes = {
  category: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
  setActiveItem: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  category: activeCategorySelector(state)
})

const mapDispatchToProps = {
  setActiveItem
}

export default connect(mapStateToProps, mapDispatchToProps)(TileGrid)
