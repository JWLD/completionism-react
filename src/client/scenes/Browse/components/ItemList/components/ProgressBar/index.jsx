import React from 'react'
import PropTypes from 'prop-types'

import * as SC from 'ProgressBar/styled'
import { BrowseBlock } from 'style/components'

const ProgressBar = props => {
  const fillPercent = props.count / props.total * 100
  const fillProp = isNaN(fillPercent) ? 0 : fillPercent

  return (
    <BrowseBlock>
      <SC.TextWrap>
        <SC.ProgressText>{props.title}</SC.ProgressText>
        <SC.ProgressText>
          {props.count} / {props.total}
        </SC.ProgressText>
      </SC.TextWrap>

      <SC.ProgressBar>
        <SC.BarFill fill={fillProp} />
      </SC.ProgressBar>
    </BrowseBlock>
  )
}

ProgressBar.propTypes = {
  count: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  total: PropTypes.number.isRequired
}

export default ProgressBar
