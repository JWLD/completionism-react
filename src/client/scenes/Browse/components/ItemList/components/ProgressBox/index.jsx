import React from 'react'
import PropTypes from 'prop-types'

import { BrowseBlock } from 'style/components'
import ProgressBar from 'ProgressBar'

import * as SC from 'ProgressBox/styled'

const ProgressBox = ({ count, title, total }) => (
  <BrowseBlock>
    <SC.TextWrap>
      <SC.ProgressText>{title}</SC.ProgressText>
      <SC.ProgressText>
        {count} / {total}
      </SC.ProgressText>
    </SC.TextWrap>

    <ProgressBar count={count} total={total} />
  </BrowseBlock>
)

ProgressBox.propTypes = {
  count: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  total: PropTypes.number.isRequired
}

export default ProgressBox
