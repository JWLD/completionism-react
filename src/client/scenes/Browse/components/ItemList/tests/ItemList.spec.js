import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import * as MOCKS from 'fixtures'
import { ItemList, mapStateToProps } from 'scenes/Browse/components/ItemList'

Enzyme.configure({ adapter: new Adapter() })

const mockProps = {
  blocks: [
    {
      name: 'Drop',
      items: [{ ...MOCKS.MOUNT_ENTRY, collected: true }]
    }
  ],
  content: 1,
  progress: {
    count: 1,
    total: 2
  },
  ...MOCKS.BROWSE_ROUTER_PROPS
}

describe('<ItemList />', () => {
  it('renders correctly', () => {
    const component = shallow(<ItemList {...mockProps} />)

    expect(component).toMatchSnapshot()
  })
})

describe('#mapStateToProps', () => {
  it('extracts the required data from state', () => {
    const result = mapStateToProps(MOCKS.INITIAL_STATE, mockProps)

    expect(Object.keys(result)).toHaveLength(3)
    expect(result).toHaveProperty('blocks')
    expect(result).toHaveProperty('content')
    expect(result).toHaveProperty('progress')
  })
})
