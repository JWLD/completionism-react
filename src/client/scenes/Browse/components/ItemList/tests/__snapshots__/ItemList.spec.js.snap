// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ItemList
    blocks={
      Array [
        Object {
          "items": Array [
            Object {
              "added": 7,
              "bmah": null,
              "class": null,
              "collected": true,
              "content": 7,
              "faction": 2,
              "icon": "inv_serpentmount_green",
              "id": 232519,
              "name": "Abyss Worm",
              "quality": 4,
              "source": 109,
              "sub1": null,
              "sub2": null,
            },
          ],
          "name": "Drop",
        },
      ]
    }
    content={1}
    progress={
      Object {
        "count": 1,
        "total": 2,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <ProgressBar
          count={1}
          title="Classic"
          total={2}
        />,
        Array [
          <ItemListBlock>
            <BlockTitle>
              Drop
            </BlockTitle>
            <ItemTile
              added={7}
              bmah={null}
              class={null}
              collected={true}
              content={7}
              faction={2}
              icon="inv_serpentmount_green"
              id={232519}
              name="Abyss Worm"
              quality={4}
              source={109}
              sub1={null}
              sub2={null}
            />
          </ItemListBlock>,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "count": 1,
          "title": "Classic",
          "total": 2,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Drop",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <BlockTitle>
              Drop
            </BlockTitle>,
            Array [
              <ItemTile
                added={7}
                bmah={null}
                class={null}
                collected={true}
                content={7}
                faction={2}
                icon="inv_serpentmount_green"
                id={232519}
                name="Abyss Worm"
                quality={4}
                source={109}
                sub1={null}
                sub2={null}
              />,
            ],
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Drop",
            },
            "ref": null,
            "rendered": "Drop",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "232519",
            "nodeType": "function",
            "props": Object {
              "added": 7,
              "bmah": null,
              "class": null,
              "collected": true,
              "content": 7,
              "faction": 2,
              "icon": "inv_serpentmount_green",
              "id": 232519,
              "name": "Abyss Worm",
              "quality": 4,
              "source": 109,
              "sub1": null,
              "sub2": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <ProgressBar
            count={1}
            title="Classic"
            total={2}
          />,
          Array [
            <ItemListBlock>
              <BlockTitle>
                Drop
              </BlockTitle>
              <ItemTile
                added={7}
                bmah={null}
                class={null}
                collected={true}
                content={7}
                faction={2}
                icon="inv_serpentmount_green"
                id={232519}
                name="Abyss Worm"
                quality={4}
                source={109}
                sub1={null}
                sub2={null}
              />
            </ItemListBlock>,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "count": 1,
            "title": "Classic",
            "total": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Drop",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <BlockTitle>
                Drop
              </BlockTitle>,
              Array [
                <ItemTile
                  added={7}
                  bmah={null}
                  class={null}
                  collected={true}
                  content={7}
                  faction={2}
                  icon="inv_serpentmount_green"
                  id={232519}
                  name="Abyss Worm"
                  quality={4}
                  source={109}
                  sub1={null}
                  sub2={null}
                />,
              ],
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Drop",
              },
              "ref": null,
              "rendered": "Drop",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "232519",
              "nodeType": "function",
              "props": Object {
                "added": 7,
                "bmah": null,
                "class": null,
                "collected": true,
                "content": 7,
                "faction": 2,
                "icon": "inv_serpentmount_green",
                "id": 232519,
                "name": "Abyss Worm",
                "quality": 4,
                "source": 109,
                "sub1": null,
                "sub2": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
